### PR TITLE
upload coverage reports to coveralls

### PR DESCRIPTION
**Pull request by pcardune**

### CODE REVIEW

**Comment by pcardune** [comment]
@Bjvanminnen the coveralls check failed because the threshold was set to 80. I put it down to 50 for now, so it won't fail for other people.

**Comment by Bjvanminnen** [comment]
Do you have any thoughts on how you expect us to use this? Is this just going to be outputting a number that we ignore, or should this change also come with a process and/or policy change?

(FWIW, my personal feelings are that code-coverage metrics are not very useful)

**Comment by pcardune** [comment]
I think the code coverage *delta* is useful information. Specifically, if a pull request decreases code coverage, that seems like something worth considering in the review process. At the moment, we don't have any tool that highlights changes to code coverage, and so it's easy to add a bunch of code with zero tests without anyone raising an eyebrow, especially when you are refactoring and it's not clear whether the refactored code is still being covered by existing tests or not.

**Comment by pcardune** [comment]
@Bjvanminnen, also, I'm curious what specifically you find to be not useful about coverage reports. Main not-useful thing I can think of is that code coverage does not guarantee that your code is bug free if your tests aren't making the right assertions, and therefore might be misleading about the quality of your tests. But as long as people understand that coverage by itself is not sufficient for having "good tests", it seems to be a very useful for figuring out what code is not tested at all.

File: apps/package.json [hunk 1]
```diff
@@ -187,6 +187,7 @@
   "dependencies": {
     "filesaver.js": "0.2.0",
     "jszip": "3.0.0",
+    "karma-coveralls": "^1.1.2",
```
Comment (bcjordan) [inline apps/package.json:191]: Curious is this where yarn adds new dependencies by default?

Comment (pcardune) [reply apps/package.json:191]: yes, just like npm.

**Comment by Bjvanminnen** [comment]
I just don't find it to be a very good metric of test quality. There are some pieces of code that are simple and straight forward enough that writing a bunch of tests isn't going to help a lot (i.e. many React components). There are other places where a small amount of code is super hairy and complex, and writing a bunch of tests would be super useful.

There is likely some correlation between the percentage of your code that is hit by tests, and whether you subjectively have good tests/good test coverage, but I don't think that this correlation is all that high. I also always worry a little bit that by introducing this as a metric, you focus on addressing the metric rather than the thing it's a proxy for (i.e. focus on trying to get that number as close to 100 as you can instead of focusing on trying to make sure we're testing the right things and testing them well).

**Comment by Bjvanminnen** [comment]
I also suspect nobody is going to look at the code coverage delta. They'll look to see if coveralls reports green or not (I suspect it only does not if you go under 50%?) and then ignore.

**Comment by pcardune** [comment]
You can make coveralls fail for changes below a given threshold. i.e. don't reduce test coverage by more than x%.

As for simple code, I think it's still important for them to be covered by tests, just not necessarily *unit* tests. For example, if you just had a react component that provided some default styling for some html, (perhaps `apps/src/templates/alert.jsx`) I probably wouldn't bother unit testing it. But presumably it's used by something else (perhaps `DisabledBubblesAlert.jsx`) that *is* worthy of tests, and then it gets covered without any extra work.

Especially in an environment where there is no static typing, it's trivially easy for a refactor to accidentally rename a variable to something that doesn't exist. All it takes is running the code and getting an undefined variable error to find this out, but if it's not covered, you'll have to wait for eyes tests to hopefully catch it before it goes out.

Anyway, I think the goal wouldn't be to make test coverage a metric, but rather just an extra piece of information to go along with code reviews. I think the ideal setup would be for the github diff itself to include the coverage information inline, so you could see quickly at review time what code is not tested at all, and make a determination then and there as to wether that is acceptable.

**Pull request merged**